spark/src/main/scala/geotrellis/spark/clip/ClipToGrid.scala

```diff
@@ -16,191 +16,111 @@
+      rdd.map(f => (mapTransform.keysForGeometry(f.geom), f))
+
+    /* Clip every Feature */
+    withKeys.flatMap { case (ks, f) => ks.flatMap(k => clipFeature(k.extent(layout), f).map((k, _))) }
+  }
+
+  /** Clips Features to fit the given Extent. Does its best to avoid unnecessary
+    * JTS interactions.
+    */
+  def byExtent[G <: Geometry, D](extent: Extent, f: Feature[G, D]): Option[Feature[Geometry, D]] = f.geom match {
+    case g: Point                              => Some(f)
+    case g: MultiPoint if coveredBy(g, extent) => Some(f)
+    case g: MultiPoint                         => clip(extent, g).map(Feature(_, f.data))
+    case g: Line       if coveredBy(g, extent) => Some(f)
+    case g: Line                               => clip(extent, g).map(Feature(_, f.data))
+    case g: MultiLine  if coveredBy(g, extent) => Some(f)
+    case g: MultiLine                          => clip(extent, g).map(Feature(_, f.data))
+    case g =>  /* Polygon and MultiPolygon */
+      val pg = PreparedGeometryFactory.prepare(g.jtsGeom)
+      val ep = extent.toPolygon.jtsGeom
+
+      if (pg.covers(ep)) Some(Feature(extent, f.data))
+      else if (pg.coveredBy(ep)) Some(f)
+      else clip(extent, g).map(Feature(_, f.data))
   }
 
-  private def clipGeom[G <: Geometry, D](
-    mapTransform: MapKeyTransform,
-    clipFeature: (Extent, Feature[G, D], Predicates) => Option[Feature[Geometry, D]],
-    feature: Feature[G, D]
-  ): Iterator[(SpatialKey, Feature[Geometry, D])] = {
-    val pointPredicates =
-      new Predicates {
-        def covers(e: Extent) = false
-        def coveredBy(e: Extent) = true
-      }
-
-    val mpOrLinePredicates =
-      new Predicates {
-        def covers(e: Extent) = false
-        def coveredBy(e: Extent) =
-          feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def polyPredicates(pg: PreparedGeometry) =
-      new Predicates {
-        def covers(e: Extent) = pg.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = pg.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def gcPredicates =
-      new Predicates {
-        def covers(e: Extent) = feature.geom.jtsGeom.covers(e.toPolygon.jtsGeom)
-        def coveredBy(e: Extent) = feature.geom.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
-      }
-
-    def clipToKey(
-      k: SpatialKey,
-      preds: Predicates
-    ): Option[(SpatialKey, Feature[Geometry, D])] = {
-      val extent = mapTransform(k)
-
-      clipFeature(extent, feature, preds).map(k -> _)
+  private def coveredBy[G <: Geometry](g: G, e: Extent): Boolean =
+    g.jtsGeom.coveredBy(e.toPolygon.jtsGeom)
+
+  private def clip[G <: Geometry](e: Extent, g: G): Option[Geometry] = {
+    val exPoly: Polygon = e.toPolygon
+
+    val clipped: Try[Geometry] = g match {
+      case mp: MultiPolygon => Try(MultiPolygon(mp.polygons.flatMap(_.intersection(exPoly).as[Polygon])))
+      case _ => Try(g.intersection(exPoly).toGeometry.get)
```

Comment:
Not throwing an Exception is a requirement of VectorPipe ingest behaviour, so we catch here. Hm, actually the catch might still be able to be done at the VP level, not here, since it will be calling `byExtent` itself in an `RDD.map`, checking for errors anyway, and logging.

Reply:
Flattening out an exception into a None is one step up from swallowing the exception: now the user knows there was something wrong, since they got a None, but have no idea what happens. I'd prefer if the error be transmitted some way, and if that's bubbling the exception up and having VP catch it to do what it needs, then that seems to work.

Reply:
Given [my suggestion above](https://github.com/locationtech/geotrellis/pull/2453#discussion_r147568159), throwing inside the `(Iterator[Extent], Feature[G, D]) => Iterator[Option[Feature[G, D]]]` still wouldn't fit VectorPipe's use-case, as we'd like to document each failed clip separately. 

How about this code here still uses `Try`, casts to `Option`, and the default GT behaviour higher up in the `clipToGrid` overloads is to detect those `Option` and throw? Returning `Try` or `Either` would be the most transparent, but you've said in the past that you prefer Exceptions because it avoids further boxing.

Either way, VectorPipe won't be using the injected `clipToGrid` methods, since it needs fine-grained control over the error reporting process and so will apply the clipping functions itself in its own `RDD.map`.

Reply:
If you turn a Try into an Option you lose all of the error information.

If VectorPipe uses it's own clip method, it seems like this shouldn't be an issue for VP. If client code wants to deal with clip-specific errors in a certain way, they should be able to do that in the passed in clip method. The current code accounts for that by having the `clipFeature` method coming in from the client return an option, so that if you wanted to handle errors by ignoring them or reporting them with some logging, you can do that and return a None.

Reply:
My plan was to exposes aliases in VP to the clipping functions in this PR, so that the user wouldn't have to think about GT imports and symbols.